pkg/supervisor/supervisor.go

```diff
@@ -21,277 +17,258 @@ import (
-				s.log.Infof("Killing pid %d", s.cmd.Process.Pid)
-				if err := s.cmd.Process.Kill(); err != nil {
-					s.log.Warnf("Failed to kill pid %d: %s", s.cmd.Process.Pid, err)
-				}
-			} else {
-				s.log.Infof("Shutting down pid %d", s.cmd.Process.Pid)
-				if err := s.cmd.Process.Signal(syscall.SIGTERM); err != nil {
-					s.log.Warnf("Failed to send SIGTERM to pid %d: %s", s.cmd.Process.Pid, err)
-				}
-			}
-			select {
-			case <-time.After(s.TimeoutStop):
-				continue
-			case <-waitresult:
-				return true
-			}
+		if err := s.maybeKillPid(); err != nil {
+			return true, fmt.Errorf("failed to kill %s process: %w", s.name, err)
 		}
+		return true, nil
 	case err := <-waitresult:
 		if err != nil {
-			s.log.Warnf("Failed to wait for process: %v", err)
-		} else {
-			s.log.Warnf("Process exited: %s", s.cmd.ProcessState)
+			s.log.Warnf("Failed to waiting for process %q: %v", s.name, err)
+			break
 		}
+		s.log.Warnf("Process %q exited: %s", s.name, s.cmd.ProcessState)
 	}
-	return false
+	return false, nil
 }
 
-// Supervise Starts supervising the given process
+// Supervise Starts supervising the given process.
 func (s *Supervisor) Supervise() error {
 	s.startStopMutex.Lock()
 	defer s.startStopMutex.Unlock()
-	// check if it is already started
 	if s.cancel != nil {
-		s.log.Warn("Already started")
+		s.log.Warn("Supervisor for %q already started", s.name)
 		return nil
 	}
-	s.log = logrus.WithField("component", s.Name)
-	s.PidFile = path.Join(s.RunDir, s.Name) + ".pid"
-
-	if err := os.MkdirAll(s.RunDir, 0755); err != nil {
+	dir := filepath.Dir(s.pidFile)
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("failed to create run dir: %w", err)
+	}
+	if err := s.maybeKillPid(); err != nil {
 		return err
 	}
+	var ctx context.Context
+	ctx, s.cancel = context.WithCancel(context.Background())
+	s.done = make(chan bool)
+	if err := s.supervise(ctx); err != nil {
+		return fmt.Errorf("failed to supervise %q: %w", s.name, err)
+	}
+	return nil
+}
 
-	if s.TimeoutStop == 0 {
-		s.TimeoutStop = 5 * time.Second
+// DetachAttr creates the proper syscall attributes to run the managed processes.
+func (s *Supervisor) detachAttr() *syscall.SysProcAttr {
+	var creds *syscall.Credential
+	if os.Geteuid() == 0 {
+		creds = &syscall.Credential{
+			Uid: uint32(s.uid),
+			Gid: uint32(s.gid),
+		}
 	}
-	if s.TimeoutRespawn == 0 {
-		s.TimeoutRespawn = 5 * time.Second
+	return &syscall.SysProcAttr{
+		Setpgid:    true,
+		Pgid:       0,
+		Credential: creds,
 	}
+}
 
-	if err := s.maybeKillPidFile(nil, nil); err != nil {
-		return err
+// supervise starts the process and waits for it to exit.
+func (s *Supervisor) supervise(ctx context.Context) error {
+	defer func() {
+		close(s.done)
+	}()
+	s.log.Infof("Starting to supervise %q", s.name)
+	s.cmd = exec.Command(s.binPath, s.args...)
+	s.cmd.Stdout = logrus.WithField("stream", "stdout").Writer()
```

Comment:
i do see this comment in logWriter

```

// logWriter implements [io.Writer] by forwarding whole lines to log. In case
// the lines get too long, it logs them in multiple chunks.
//
// This is in contrast to logrus's implementation of io.Writer, which simply
// errors out if the log line gets longer than 64k.
```

Reply:
I have tested this and it works just fine with lines longer than 64k, that is why I removed it. Maybe that was a problem but got fixed in newer logrus versions? I don't know. Anyways, it logs just fine.

Reply:
For sake of completeness: I tested with log lines > 1MB.